### ISSUE TITLE
Popup for "Error Contacting Site" should offer to delete the update site

### ISSUE THREAD

**Issue by erik-brangs** [eclipse-equinox/p2#207]
When searching for new items to install and when an update site cannot be contacted, there is a popup "Error contacting site" that says "No software site found at $URL. Do you wish to edit that location?".

When this occurs during an installation that contacts all update sites, this popup is not very helpful. It's difficult to remember the affected site, particularly if this occurs for multiple sites. Additionally, this may affect older update sites that aren't used any longer. It would be useful to give the user the option to simply delete outdated sites so that they don't cause problems in the future. Another useful possiblity would be to allow the user to disable the location. That would make the outdated site easier to find in the software sites list later.

**Comment by mickaelistria** [comment]
That's a good proposal. Would you be interested in trying to submit a PR implementing such enhancement?

**Comment by erik-brangs** [comment]
Yes, in theory. However, I'm not sure when I will get around to do it. So it's fine for me if someone else wants to do it instead.

**Comment by ajohnson1** [comment]
A related idea is whether a 'HTTP/1.1 301 Moved Permanently' response should update the location - automatically or with a 'Yes' 'No' and perhaps one of 'Never' (meaning for the update site source location) or 'Never' (meaning for the update site source location and redirection location pair).
https://www.rfc-editor.org/rfc/rfc2616#section-10.3.2
>The requested resource has been assigned a new permanent URI and any
   future references to this resource SHOULD use one of the returned
   URIs.  Clients with link editing capabilities ought to automatically
   re-link references to the Request-URI to one or more of the new
   references returned by the server, where possible. This response is
   cacheable unless indicated otherwise.
 
https://www.rfc-editor.org/rfc/rfc7231#section-6.4.2

> The 301 (Moved Permanently) status code indicates that the target
   resource has been assigned a new permanent URI and any future
   references to this resource ought to use one of the enclosed URIs.
   Clients with link-editing capabilities ought to automatically re-link
   references to the effective request URI to one or more of the new
   references sent by the server, where possible.
   
   appear to suggest that it should happen automatically.
   
Also, a 410 Gone should probably delete the location entry (and leave the name) or disable the update site (or remove the whole update site?). Is it useful for a user to see that that there used to be a software site but it isn't now usable.
https://www.rfc-editor.org/rfc/rfc7231#section-6.5.9

>The 410 response is primarily intended to assist the task of web
   maintenance by notifying the recipient that the resource is
   intentionally unavailable and that the server owners desire that
   remote links to that resource be removed.  Such an event is common
   for limited-time, promotional services and for resources belonging to
   individuals no longer associated with the origin server's site.

**Comment by HannesWell** [comment]
> A related idea is whether a 'HTTP/1.1 301 Moved Permanently' response should update the location - automatically or with a 'Yes' 'No' and perhaps one of 'Never' (meaning for the update site source location) or 'Never' (meaning for the update site source location and redirection location pair). https://www.rfc-editor.org/rfc/rfc2616#section-10.3.2
> 

Sounds like a very hand improvement to me. The 410 would also help to distinguish between an intended removal from the provider of the repo and a potential network issue.
In general before a repo/site is removed the user should be asked at least once. The options could be Yes/No/Always/Never. The latter two should be persisted in a preference so that you can edit that choice or for example set it in an Oomph setup.

**Comment by laeubi** [comment]
Just before getting to enthusiastic about the technical aspects I'd like to note that P2 do not handle anything in this regard but ECF does it, so for P2 there is no such thing as HTTP status codes or anything else available. Then on the UI-side again don't know much about the transport-failure at all as it just tries to load a URI... So I think the very first thing should be stupid simple that when an error occurs asking the user to remove that location.

**Comment by ajohnson1** [comment]
From a look at the code I think sometimes HTTP status codes are available but perhaps not for redirects.

https://github.com/eclipse-equinox/p2/blob/23e53212405d35827302deb2913616b2a3e637eb/bundles/org.eclipse.equinox.p2.transport.ecf/src/org/eclipse/equinox/internal/p2/transport/ecf/RepositoryStatus.java#L121

BrowseFileTransferException
https://www.eclipse.org/ecf/org.eclipse.ecf.docs/api/org/eclipse/ecf/filetransfer/BrowseFileTransferException.html#getErrorCode()
IncomingFileTransferException
https://www.eclipse.org/ecf/org.eclipse.ecf.docs/api/org/eclipse/ecf/filetransfer/IncomingFileTransferException.html
